### PR TITLE
usb: Work on hubs

### PR DESCRIPTION
**Pull request by alexkalmuk**
* Add usb hubs (working with ohci on qemu)
* Make usb core and api simpler

### CODE REVIEW

File: scripts/qemu/auto_qemu_with_usb_storage [hunk 1]
```diff
@@ -7,9 +7,11 @@ then
 	echo "$image_file_name found."
 else
 	echo "$image_file_name not found. A new one will be created now..."
+	mkfs -t vfat $image_file_name
```
Comment (Deryugin) [inline scripts/qemu/auto_qemu_with_usb_storage]: Should this line go after `qemu-img create ...`? If file doesn't exist, mkfs will fail

File: src/drivers/usb/hc/ehci_hcd.c [hunk 2]
```diff
@@ -576,50 +603,77 @@ static void ehci_qh_fill(struct ehci_hcd *ehci, struct usb_request *req, struct
...
-		if (req->endp->hci_specific) {
-			new_qh = req->endp->hci_specific;
+		/* Data stage (optional) */
+		if (req->len > 0) {
+			if (req->token & USB_TOKEN_OUT)  {
+				token = EHCI_QTD_PID_OUT << 8;
+			} else {
+				token = EHCI_QTD_PID_IN << 8;
+			}
+			token |= EHCI_QTD_STS_ACTIVE | (3 << 10);
```
Comment (Deryugin) [inline src/drivers/usb/hc/ehci_hcd.c]: I think it's better to replace `3 << 10` with some macro

File: src/drivers/usb/class/usb_mass_storage.c [hunk 3]
```diff
@@ -152,40 +125,47 @@ static void usb_mass_handle(struct usb_class *cls, struct usb_dev *dev) {
...
 			&mass->maxlun);
 }
 
-static void usb_mass_release(struct usb_class *cls, struct usb_dev *dev) {
-	struct usb_mass *mass = usb2massdata(dev);
+static int usb_ms_probe(struct usb_dev *dev) {
+	struct usb_mass *mass = pool_alloc(&usb_mass_classes);
 
-	scsi_dev_detached(&mass->scsi_dev);
+	mass = pool_alloc(&usb_mass_classes);
```
Comment (Deryugin) [inline src/drivers/usb/class/usb_mass_storage.c]: Double pool allocation (here and 2 lines upper)

File: src/drivers/usb/usb_config.c [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+		goto err;
+	}
+
+	ret = usb_endp_control_wait(dev->endpoints[0],
+		USB_DIR_IN | USB_REQ_TYPE_STANDARD | USB_REQ_RECIP_DEVICE,
+		USB_REQ_GET_DESCRIPTOR,
+		USB_DESC_TYPE_CONFIG << 8,
+		n, len, dev->config_buf, 1000);
+	if (ret < 0) {
+		goto err;
```
Comment (Deryugin) [inline src/drivers/usb/usb_config.c]: `sysfree(dev->config_buf)` to avoid memory leak

File: scripts/qemu/auto_qemu_with_usb_storage (hunk 1 above)
Comment (alexkalmuk) [reply scripts/qemu/auto_qemu_with_usb_storage]: Yes, thank you, fixed.

File: src/drivers/usb/hc/ehci_hcd.c (hunk 2 above)
Comment (alexkalmuk) [reply src/drivers/usb/hc/ehci_hcd.c]: I added `3 << EHCI_QTD_CERR_SHIFT` instead.

File: src/drivers/usb/class/usb_mass_storage.c (hunk 3 above)
Comment (alexkalmuk) [reply src/drivers/usb/class/usb_mass_storage.c]: Thank you, fixed.

File: src/drivers/usb/usb_config.c (hunk 4 above)
Comment (alexkalmuk) [reply src/drivers/usb/usb_config.c]: Fixed.

**Pull request merged**